File: frontend/src/api/ai.ts
import request from '@/utils/request'

export interface AIChatRequest {
  prompt: string
  system_prompt?: string
  temperature?: number
  max_tokens?: number
}

export interface AIChatResponse {
  content: string
  model: string
}

export interface AIScoreRequest {
  question: string
  correct_answers: string[]
  user_answer: string
}

export interface AIScoreResponse {
  score: number
  accuracy_score: number
  artistic_score: number
  diction_score: number
  feedback: string
  is_correct: boolean
}

export interface AIReferencePoemInfo {
  title: string
  author: string
  dynasty: string
  content: string
  genre?: string
}

export interface AICreationRequest {
  context: string
  mode: 'continue' | 'inspire' | 'generate' | 'theme' | 'imitate_guide'
  keywords?: string[]
  reference_poem?: AIReferencePoemInfo
}

export interface AICreationResponse {
  content: string
  explanation: string
  suggestions?: string[]
}

export interface AICheckPoemRequest {
  poem_text: string
}

export interface AICheckPoemIssue {
  type: string
  location: string
  description: string
}

export interface AICheckPoemResponse {
  is_valid: boolean
  tone_analysis?: any[]
  rhyme_analysis?: Record<string, any>
  couplet_analysis?: Record<string, any>
  issues: AICheckPoemIssue[]
  suggestions: string[]
}

export interface AIAnalyzePoemRequest {
  poem_text: string
}

export interface AIAnalyzePoemResponse {
  total_score: number
  meter_score: number
  artistic_score: number
  diction_score: number
  overall_score: number
  highlights: string[]
  improvements: string[]
  appreciation: string
}

export function aiChat(data: AIChatRequest) {
  return request.post<AIChatResponse>('/ai/chat', data)
}

export function aiScore(data: AIScoreRequest) {
  return request.post<AIScoreResponse>('/ai/score', data)
}

const AI_TIMEOUT = { timeout: 60000 }
const AI_CONTEXT_TIMEOUT = { timeout: 120000 }

export function aiCreation(data: AICreationRequest) {
  return request.post<AICreationResponse>('/ai/creation', data, AI_TIMEOUT)
}

export function aiCheckPoem(data: AICheckPoemRequest) {
  return request.post<AICheckPoemResponse>('/ai/check-poem', data, AI_TIMEOUT)
}

export function aiAnalyzePoem(data: AIAnalyzePoemRequest) {
  return request.post<AIAnalyzePoemResponse>('/ai/analyze-poem', data, AI_TIMEOUT)
}

export type PoemContextQueryType = 'author_bio' | 'deep_appreciation' | 'allusions' | 'verse_analysis' | 'free_qa' | 'meter_analysis'

export interface AIPoemContextRequest {
  title: string
  author: string
  dynasty: string
  content: string
  genre?: string
  category?: string
  query_type: PoemContextQueryType
  question?: string
}

export interface AIPoemContextSection {
  heading: string
  text: string
}

export interface AIMeterLine {
  text: string
  tones: string[]
  rhyme: string
  couplet: string
  note: string
}

export interface AIPoemContextResponse {
  query_type: string
  content: string
  title?: string
  sections?: AIPoemContextSection[]
  lines?: AIMeterLine[]
  rhyme_scheme?: string
  meter_type?: string
}

export function aiPoemContext(data: AIPoemContextRequest) {
  return request.post<AIPoemContextResponse>('/ai/poem-context', data, AI_CONTEXT_TIMEOUT)
}

export interface AIPoemChatMessage {
  role: 'user' | 'assistant'
  content: string
}

export interface AIPoemChatRequest {
  title: string
  author: string
  dynasty: string
  content: string
  genre?: string
  category?: string
  history: AIPoemChatMessage[]
  message: string
}

export interface StreamEvent {
  type: 'content' | 'thinking' | 'tool_call' | 'memory' | 'error' | 'done'
  content?: string
  tools?: string[]
  labels?: string[]
  round?: number
}

export async function aiPoemChatStream(
  data: AIPoemChatRequest,
  onChunk: (text: string) => void,
  onDone: () => void,
  onError?: (err: string) => void,
  onEvent?: (event: StreamEvent) => void,
) {
  const baseURL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1'
  const userStore = (await import('@/store/modules/user')).useUserStore()
  const token = userStore.token

  const response = await fetch(`${baseURL}/ai/poem-chat-stream`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify(data),
  })

  if (!response.ok || !response.body) {
    onError?.('AI对话服务暂时不可用')
    return
  }

  const reader = response.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  while (true) {
    const { done, value } = await reader.read()
    if (done) break

    buffer += decoder.decode(value, { stream: true })
    const lines = buffer.split('\n')
    buffer = lines.pop() || ''

    for (const line of lines) {
      if (!line.startsWith('data: ')) continue
      const payload = line.slice(6).trim()
      if (payload === '[DONE]') {
        onDone()
        return
      }
      try {
        const parsed = JSON.parse(payload)
        if (parsed.error) {
          onError?.(parsed.error)
          return
        }
        if (parsed.type === 'thinking' || parsed.type === 'tool_call' || parsed.type === 'memory') {
          onEvent?.(parsed as StreamEvent)
        } else if (parsed.content) {
          onChunk(parsed.content)
        }
      } catch {
      }
    }
  }
  onDone()
}

export async function aiStreamRequest<T>(
  url: string,
  data: Record<string, any>,
  onEvent: (event: StreamEvent) => void,
  onResult: (result: T) => void,
  onError?: (err: string) => void,
) {
  const baseURL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1'
  const userStore = (await import('@/store/modules/user')).useUserStore()
  const token = userStore.token

  const response = await fetch(`${baseURL}${url}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify(data),
  })

  if (!response.ok || !response.body) {
    onError?.('AI服务暂时不可用')
    return
  }

  const reader = response.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  while (true) {
    const { done, value } = await reader.read()
    if (done) break

    buffer += decoder.decode(value, { stream: true })
    const lines = buffer.split('\n')
    buffer = lines.pop() || ''

    for (const line of lines) {
      if (!line.startsWith('data: ')) continue
      const payload = line.slice(6).trim()
      if (!payload) continue
      try {
        const parsed = JSON.parse(payload)
        if (parsed.type === 'error') {
          onError?.(parsed.content || 'AI服务暂时不可用')
          return
        }
        if (parsed.type === 'result') {
          onResult(parsed.data as T)
          return
        }
        if (parsed.type === 'thinking' || parsed.type === 'tool_call') {
          onEvent(parsed as StreamEvent)
        }
      } catch {}
    }
  }
}

export function aiCheckPoemStream(
  data: AICheckPoemRequest,
  onEvent: (event: StreamEvent) => void,
  onResult: (result: AICheckPoemResponse) => void,
  onError?: (err: string) => void,
) {
  return aiStreamRequest<AICheckPoemResponse>('/ai/check-poem-stream', data, onEvent, onResult, onError)
}

export function aiAnalyzePoemStream(
  data: AIAnalyzePoemRequest,
  onEvent: (event: StreamEvent) => void,
  onResult: (result: AIAnalyzePoemResponse) => void,
  onError?: (err: string) => void,
) {
  return aiStreamRequest<AIAnalyzePoemResponse>('/ai/analyze-poem-stream', data, onEvent, onResult, onError)
}

export function aiCreationStream(
  data: AICreationRequest,
  onEvent: (event: StreamEvent) => void,
  onResult: (result: AICreationResponse) => void,
  onError?: (err: string) => void,
) {
  return aiStreamRequest<AICreationResponse>('/ai/creation-stream', data, onEvent, onResult, onError)
}

export function aiPoemContextStream(
  data: AIPoemContextRequest,
  onEvent: (event: StreamEvent) => void,
  onResult: (result: AIPoemContextResponse) => void,
  onError?: (err: string) => void,
) {
  return aiStreamRequest<AIPoemContextResponse>('/ai/poem-context-stream', data, onEvent, onResult, onError)
}

File: backend/app/schemas/ai.py
from typing import Optional
from pydantic import BaseModel, Field


class AIChatRequest(BaseModel):
    prompt: str = Field(..., min_length=1, max_length=5000)
    system_prompt: Optional[str] = Field(None, max_length=2000)
    temperature: float = Field(0.7, ge=0, le=2)
    max_tokens: int = Field(2048, ge=1, le=8192)


class AIChatResponse(BaseModel):
    content: str
    model: str


class AIScoreRequest(BaseModel):
    question: str = Field(..., min_length=1)
    correct_answers: list[str] = Field(..., min_length=1)
    user_answer: str = Field(..., min_length=1)


class AIScoreResponse(BaseModel):
    score: int
    accuracy_score: int
    artistic_score: int
    diction_score: int
    feedback: str
    is_correct: bool


class AIReferencePoemInfo(BaseModel):
    title: str = Field(..., max_length=200)
    author: str = Field(..., max_length=100)
    dynasty: str = Field(..., max_length=50)
    content: str = Field(..., max_length=5000)
    genre: Optional[str] = None


class AICreationRequest(BaseModel):
    context: str = Field(..., min_length=1, max_length=2000)
    mode: str = Field("continue", pattern="^(continue|inspire|generate|theme|imitate_guide)$")
    keywords: Optional[list[str]] = None
    reference_poem: Optional[AIReferencePoemInfo] = None


class AICreationResponse(BaseModel):
    content: str
    explanation: str
    suggestions: Optional[list[str]] = None


class AICheckPoemRequest(BaseModel):
    poem_text: str = Field(..., min_length=1, max_length=2000)


class AICheckPoemResponse(BaseModel):
    is_valid: bool
    tone_analysis: Optional[list] = None
    rhyme_analysis: Optional[dict] = None
    couplet_analysis: Optional[dict] = None
    issues: list = Field(default_factory=list)
    suggestions: list = Field(default_factory=list)


class AIAnalyzePoemRequest(BaseModel):
    poem_text: str = Field(..., min_length=1, max_length=2000)


class AIAnalyzePoemResponse(BaseModel):
    total_score: int
    meter_score: int
    artistic_score: int
    diction_score: int
    overall_score: int
    highlights: list[str] = Field(default_factory=list)
    improvements: list[str] = Field(default_factory=list)
    appreciation: str


class AIFeihuaRequest(BaseModel):
    keyword: str = Field(..., min_length=1, max_length=4)
    used_poems: list[str] = Field(default_factory=list)


class AIFeihuaResponse(BaseModel):
    verse: str
    poem_title: str
    author: str
    dynasty: str
    confidence: float


class AIPoemContextRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    author: str = Field(..., min_length=1, max_length=100)
    dynasty: str = Field(..., min_length=1, max_length=50)
    content: str = Field(..., min_length=1, max_length=5000)
    genre: Optional[str] = None
    category: Optional[str] = None
    query_type: str = Field(
        ...,
        pattern="^(author_bio|deep_appreciation|allusions|verse_analysis|free_qa|meter_analysis)$",
    )
    question: Optional[str] = Field(None, max_length=500)


class AIPoemContextResponse(BaseModel):
    query_type: str
    content: str
    title: Optional[str] = None
    sections: Optional[list[dict]] = None
    lines: Optional[list[dict]] = None
    rhyme_scheme: Optional[str] = None
    meter_type: Optional[str] = None


class AIPoemChatMessage(BaseModel):
    role: str = Field(..., pattern="^(user|assistant)$")
    content: str = Field(..., min_length=1, max_length=2000)


class AIPoemChatRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    author: str = Field(..., min_length=1, max_length=100)
    dynasty: str = Field(..., min_length=1, max_length=50)
    content: str = Field(..., min_length=1, max_length=5000)
    genre: Optional[str] = None
    category: Optional[str] = None
    history: list[AIPoemChatMessage] = Field(default_factory=list, max_length=20)
    message: str = Field(..., min_length=1, max_length=500)


class AIPoemChatResponse(BaseModel):
    reply: str
    title: Optional[str] = None
